### PR TITLE
[ONNX] Changes to export API to better handle named arguments

### PR DESCRIPTION
**Pull request by shubhambhokare1**
The args parameter of ONNX export is changed to better support optional arguments such that args is represented as: 
args (tuple of arguments or torch.Tensor, a dictionary consisting of named arguments (optional)): 
            a dictionary to specify the input to the corresponding named parameter: 
            - KEY: str, named parameter 
            - VALUE: corresponding input

### CODE REVIEW

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 1]
```diff
@@ -81,14 +81,18 @@ def run_model_test(self, model, batch_size=2, state_dict=None,
...
     with torch.no_grad():
         if isinstance(input, torch.Tensor):
             input = (input,)
         # In-place operators will update input tensor data as well.
         # Thus inputs are replicated before every forward call.
-        input_copy = copy.deepcopy(input)
-        output = model(*input_copy)
+        input_args = copy.deepcopy(input)
+        input_kwargs = {}
+        print(input_args)
```
Comment (jiafatom) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:91]: Do we need this print?

File: torch/onnx/utils.py [hunk 2]
```diff
@@ -316,6 +316,18 @@ def _decide_external_data_format(use_external_data_format, operator_export_type,
     model_file_location = f if val_use_external_data_format and isinstance(f, str) else str()
     return val_use_external_data_format, model_file_location
 
+def _decide_input_format(model, args):
+    import inspect
+    sig = inspect.signature(model.forward)
+    ordered_list_keys = list(sig.parameters.keys())
+    if isinstance(args[-1], dict):
```
Comment (jiafatom) [inline torch/onnx/utils.py:323]: This assumes that the dictionary input must come at last, shall we document it somewhere?

Comment (shubhambhokare1) [reply torch/onnx/utils.py:323]: Adding documentation

**Comment by BowenBao** [comment]
fix CI test failures?

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 3]
```diff
@@ -470,7 +475,49 @@ def forward(self, x_in):
...
+                if y is not None:
+                    return x + y
+                if z is not None:
+                    return x + z
+                return x
+
+        x = torch.randn(2, 3)
+        y = torch.randn(2, 3)
+        # Without optional arguments dictionary
+        self.run_test(MixedModel(), (x, y, None))
```
Comment (neginraoof) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:504]: What if y is None and z is a tensor? Would this work?
 self.run_test(MixedModel(), (x, None, z))

For the named input case, I guess this should work. Can we add a test:
self.run_test(MixedModel(), (x, {'y': None, 'z': z}))

Comment (shubhambhokare1) [reply test/onnx/test_pytorch_onnx_onnxruntime.py:504]: Both these cases work with the current implementation, added both tests as well

**Comment by BowenBao** [comment]
Can you update the comments/documentation in `__init__.py`?

File: docs/source/onnx.rst [hunk 4]
```diff
@@ -274,6 +274,70 @@ In addition, Dropout layer need defined in init function so that inferencing can
...
+  is mandatory to have an empty dictionary as the last argument in the args tuple. 
+  For example, ::
+
+      class Model(torch.nn.Module): 
+        def forward(self, k, x): 
+          ...  
+          return x 
+      m = Model() 
+      k = torch.randn(2, 3)   
+      x = {torch.tensor(1.): torch.randn(2, 3)}
```
Comment (BowenBao) [inline docs/source/onnx.rst:330]: is this line correct?

Comment (shubhambhokare1) [reply docs/source/onnx.rst:330]: Similar to tests test_dict_str

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 5]
```diff
@@ -470,7 +475,101 @@ def forward(self, x_in):
...
+        self.run_test(NoOptionalModel(), (x,))
+        # With empty optional arguments dictionary
+        y = torch.randn(2, 3)
+        self.run_test(NoOptionalModel(), (y, {}))
+
+    def test_optional_inputs_with_mixed_optionals(self):
+        class MixedModel(torch.nn.Module):
+            def forward(self, x, y=None, z=None):
+                if y is not None:
+                    return x + y
```
Comment (BowenBao) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:496]: nit: it would be a better test case if y branch and z branch are not the same ..

**Review by BowenBao — APPROVED** [review]
LGTM, minor comment inline.

**Comment by shubhambhokare1** [comment]
@bzinodev Anything else required on my end for this to be merged?

**Review by bzinodev — APPROVED** [review]

**Pull request closed without merging**